gitpod-io/gitpod
# [server] Reduce messagebus load (drastically)

## Description
This fixes our current issues with "too many messages queues".

After investigating potential leaks in `server` there were no obvious ones. But it turns out that just create a ton of websocket connections. And each websocket connection creates `n` queues (where `n` [increased substantially lately with the introduction of T&P](https://github.com/gitpod-io/gitpod/blob/d4ecb355aada111d91e12c6b13a8470e65c37bf2/components/server/ee/src/workspace/gitpod-server-impl.ts#L85-L93)) ([details](https://github.com/gitpod-io/gitpod/blob/d4ecb355aada111d91e12c6b13a8470e65c37bf2/components/server/src/messaging/local-message-broker.ts#L43-L56)). 

The [pattern we see in production](https://github.com/gitpod-io/gitpod/issues/6662#issuecomment-968858687) regularly is thus on "us leaking connections/queues", but more "users leaking tabs". :grimacing: 
While we should also look into reducing the number of websocket connections from dashboard, there are some we cannot reduce (2 per workspace, for instance). Instead, this PR attempts to reduce the factor of "queues per connection" from `n` to `0` by:
 - establishing only one messagebus listener per server instance per topic (listening to _all_ messages)
 - doing the message [filtering and delivery in `server`](https://github.com/gitpod-io/gitpod/blob/d4ecb355aada111d91e12c6b13a8470e65c37bf2/components/server/src/messaging/local-message-broker.ts#L72-L126) (turns out to be simple)

This has the benefits that:
 - it clearly shows how much of the rabbitmq features we use (none)
 - makes it easier to swap out rabbitmq with something else
 - is a "drop-in" replacement for the current code, so we do not have to verify every dashboard page to still receive the correct messages (nr and order of messages stays identical)

Credits to @AlexTugarev for the idea and discussions! :pray: 

## Related Issue(s)

Fixes #6662.

## How to test

 1. start a workspace, open & close ports, and stop it
 2. notice how the dashboards updates as expected

## Release Notes

```release-note
NONE
```

## Documentation
